server/src/document/document.controller.ts

```diff
@@ -65,4 +76,80 @@ export class DocumentController {
       response.status(400).send({ "error": 'You can only upload files to dcp_communityboarddisposition at this time' });
     }
   }
+
+  // "path" refers to the "relative server path", the path
+  // to the file itself on the sharepoint host. It has the format
+  //
+  //   /sites/<site>/<entity_name>/<folder_name>/filename.pdf
+  //
+  // For example,
+  //
+  //   /sites/dcpuat2/dcp_package/2021M0268_DraftEAS_1_996699F37323EB11A813001DD8309FA8/EAS%20Full%20Form.pdf
+  //
+  // Note that <folder_name> has this format:
+  //
+  //   <package_name>_<PACKAGEID>
+  //
+  // where package_name is the dcp_name property on the package
+  // (spaces replaced with underscores)
+  // and PACKAGEID is the package dcp_packageid property, stripped
+  // of hyphens.
+  //
+  // We have assurance of this. See
+  // https://dcp-paperless.visualstudio.com/dcp-paperless-dynamics/_workitems/edit/13366
+  @Get('/*')
+  async read(@Param() path, @Res() res) {
+    const pathSegment = path[0];
+    const packageFolder = pathSegment.split('/')[3];
+    const packageFolderSegments = packageFolder.split('_');
+    const strippedPackageId = packageFolderSegments[packageFolderSegments.length - 1];
+
+    // we need to re-insert hyphens to recreate the actual packageId
+    // TODO: Consider asking to preseve hyphens in Package ID
+    const packageId = [
+      strippedPackageId.slice(0, 8),
+      '-',
+      strippedPackageId.slice(8, 12),
+      '-',
+      strippedPackageId.slice(12,16),
+      '-',
+      strippedPackageId.slice(16,20),
+      '-',
+      strippedPackageId.slice(20),
+    ].join('');
+
+    try {
+      // Only documents belonging to public, submitted packages should be accessible
+      // to the public. So we make sure the requested document is associated with at least one
+      // public, submitted package.
+      const { records: [firstPackage] } = await this.crmService.get('dcp_packages', `
+        $select=dcp_packageid
+        &$filter=dcp_packageid eq ${packageId}
+        and (
+          dcp_visibility eq ${PACKAGE_VISIBILITY.GENERAL_PUBLIC}
+        )
+        and (
+          statuscode eq ${PACKAGE_STATUSCODE.SUBMITTED}
+        )
+      `);
+
+      if (!firstPackage) {
+        const errorMessage = `No document access.`;
+        throw new HttpException({
+          "code": "NO_DOCUMENT_ACCESS",
+          "detail": errorMessage,
+        }, HttpStatus.FORBIDDEN);
+      }
+
+      const stream = await this.crmService.getSharepointFile(pathSegment);
+
+      stream.pipe(res);
+    } catch(e) {
+      const errorMessage = `Unable to provide document access.`;
+      throw new HttpException({
+        "code": "DOCUMENT_GET_ERROR",
+        "detail": errorMessage,
+      }, HttpStatus.FORBIDDEN);
+    }
```

Comment:
i would look critically at how errors have been working for us in applicant portal and how we should improve on that. I have personally found it difficult to debug errors because the actual error message has been missing or because the error message falsely describes what actually happeened.

Reply:
My original thought is to, on principle, hide those CRM error details to the client/user and only log it in the backend (I should have included console.logs here). Though I think it in our case since it's an alpha/beta app and there's such high value for us to directly see the CRM errors in the client, that would make sense to include `e.message` in the exception payload.  

I'll include it here to get us started.
I'll create a task for us to work out Zap Search frontend/backend error reporting